src/selectors_vm/stack.rs

```diff
@@ -52,13 +52,96 @@ impl ChildCounter {
     }
 }
 
+struct CounterItem {
+    /// The counter at this index
+    pub counter: ChildCounter,
+    /// The index of this counter in the stack
+    pub index: usize,
+}
+
+struct CounterList {
+    items: Vec<CounterItem>,
+    // we always have at least one item, an empty list shouldn't exist in the map
+    current: CounterItem,
+}
+
+#[derive(Default)]
+/// A more efficient counter that only requires one owned local name to track counters across multiple stack frames
+pub struct TypedChildCounterMap(HashMap<LocalName<'static>, CounterList>);
+
+impl TypedChildCounterMap {
+    /// Adds a seen child to the map. The index is the level of the item
+    pub fn add_child(&mut self, name: &LocalName, index: usize) {
+        let hash = {
```

Comment:
Nit: move to a separate function instead of closure

Reply:
It's an expression (a closure requires `||`), but sure.